third_party/move/move-compiler/src/typing/expand.rs

```diff
@@ -57,21 +57,65 @@ fn types(context: &mut Context, ss: &mut Vec<Type>) {
     }
 }
 
-pub fn type_(context: &mut Context, ty: &mut Type) {
+// type_ for expanding the `i`-th type param of struct `ty_name`
+fn type_struct_ty_param(context: &mut Context, ty: &mut Type, i: usize, ty_name: &TypeName_) {
+    match ty_name {
+        TypeName_::ModuleType(mod_id, struct_name) => {
+            let param_name = &context
+                .struct_definition(mod_id, struct_name)
+                .type_parameters[i]
+                .param
+                .user_specified_name
+                .value;
+            let msg = format!("Cannot infer the type parameter `{param_name}` for generic struct `{ty_name}`. Try providing a type parameter.");
+            type_msg_(context, ty, &msg);
+        },
+        _ => type_(context, ty),
+    }
+}
+
+fn type_fun_ty_param(
+    context: &mut Context,
+    ty: &mut Type,
+    i: usize,
+    m: &ModuleIdent,
+    n: &FunctionName,
+) {
+    let param_name = context.function_info(m, n).signature.type_parameters[i].user_specified_name;
+    type_msg_(context, ty, &format!("Cannot infer the type parameter `{param_name}` for generic function `{m}::{n}`. Try providing a type parameter."));
+}
+
+// Try to expand the type of ty, warning with msg_uninferred if type cannot be inferred.
+fn type_msg_(context: &mut Context, ty: &mut Type, msg_uninferred: &str) {
     use Type_::*;
     match &mut ty.value {
         Anything | UnresolvedError | Param(_) | Unit => (),
         Ref(_, b) => type_(context, b),
         Var(tvar) => {
             let ty_tvar = sp(ty.loc, Var(*tvar));
-            let replacement = core::unfold_type(&context.subst, ty_tvar);
+            fn unfold_type_mut(subst: &mut Subst, sp!(loc, t_): Type) -> Type {
```

Comment:
Why is `core::unfold_type` not doing what you are trying to achieve here? It should unfold the type by its final replacement, does it not?